stdlib/public/core/Sequence.swift

```diff
@@ -355,6 +355,31 @@ public protocol Sequence<Element> {
   ///   In this case, see the documentation of `Collection.underestimatedCount`.
   var underestimatedCount: Int { get }
 
+  /// Sequences whose `Element` is `Equatable` and that are able to quickly
+  /// check if they contain a particular value can implement this requirement
+  /// to speed up the standard `contains` method.
+  ///
+  /// The default implementation returns nil, indicating that `contains` should
+  /// fall back to the standard linear search algorithm.
+  ///
+  /// `Sequence` and `Collection` algorithms other than `contains` itself may
+  /// adapt their behavior based on whether or not this function returns nil.
+  /// For example, a generic algorithm that needs to do containment checks for
+  /// many different values may decide not to copy items into a temporary `Set`
+  /// if it sees that the sequence implements this method. Therefore, sequences
+  /// should only implement this method if they can do it in better than linear
+  /// time.
+  ///
+  /// This method must not return non-nil on sequences that are destructively
+  /// consumed by iteration.
```

Comment:
Is this a safe assumption/requirement to make? It seems like it should be, but this hasn't been communicated before. (Trying to think of a consuming sequence that could implement this in sub-linear time and not having any luck.)

Reply:
Maybe we should instead say that this method does not consume any elements:

```suggestion
  /// For sequences that are destructively consumed by iteration, calling this
  /// method must not consume any elements. (Such sequences usually leave this
  /// method with its default, `nil`-returning implementation, which trivially
  /// satisfies this requirement.)
```